Branch-SDK/src/main/java/io/branch/referral/Branch.java

```diff
@@ -2555,10 +2554,25 @@ public void onAppLinkFetchFinished(String nativeAppLinkUrl) {
             }
         }
 
-        if (getInitState() == SESSION_STATE.UNINITIALISED || !isFirstInitialization) {
+        // Check if new session is being forced. There are two use cases for setting the ForceNewBranchSession to true:
+        // 1. Launch an activity via a push notification while app is in foreground but does not have
+        // the particular activity in the backstack, in such cases, users can't utilize reInitSession() because
+        // it's called from onNewIntent() which is never invoked
+        // todo: this is tricky for users, get rid of ForceNewBranchSession if possible. (if flag is not set, the content from Branch link is lost)
+        // 2. Some users navigate their apps via Branch links so they would have to set ForceNewBranchSession to true
+        // which will blow up the session count in analytics but does the job.
+        Intent intent = currentActivityReference_ != null && currentActivityReference_.get() != null ?
+                currentActivityReference_.get().getIntent() : null;
+        boolean forceBranchSession = checkIntentForSessionRestart(intent);
+
+        boolean reInitSession = !isFirstInitialization;// use case: launch current (or former but currently partially visible) activity via a push notification
+
+        if (getInitState() == SESSION_STATE.UNINITIALISED || reInitSession | forceBranchSession) {
```

Comment:
* Watch for bitwise OR
* Just use !isFirstInitialization variable that we already have

Reply:
Fixed.